vendors/st/boards/stm32l475_discovery/ports/common_io/iot_spi.c

```diff
@@ -0,0 +1,679 @@
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_spi_transfer_async( IotSPIHandle_t const pxSPIPeripheral,
+                                uint8_t * const pvTxBuffer,
+                                uint8_t * const pvRxBuffer,
+                                size_t xBytes )
+{
+    int32_t lError = IOT_SPI_SUCCESS;
+
+    /* TODO: Add case for slave not selected? */
+    if( ( pxSPIPeripheral == NULL ) || ( pxSPIPeripheral->sOpened == IOT_SPI_CLOSED ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else if( ( pvTxBuffer == NULL ) || ( pvRxBuffer == NULL ) || ( xBytes == 0 ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        if( HAL_SPI_GetState( pxSpi ) == HAL_SPI_STATE_BUSY_TX_RX )
+        {
+            lError = IOT_SPI_BUS_BUSY;
+        }
+        else
+        {
+            HAL_NVIC_SetPriority( ( IRQn_Type ) ( ( STM32_SPI_HalContext_t * ) pxSPIPeripheral->pxSpiContext )->eIrqNum, 1, 0 );
+            HAL_NVIC_EnableIRQ( ( IRQn_Type ) ( ( STM32_SPI_HalContext_t * ) pxSPIPeripheral->pxSpiContext )->eIrqNum );
+
+            if( HAL_SPI_TransmitReceive_IT( pxSpi, pvTxBuffer, pvRxBuffer, ( uint16_t ) xBytes ) != HAL_OK )
+            {
+                lError = IOT_SPI_TRANSFER_ERROR;
+            }
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_spi_close( IotSPIHandle_t const pxSPIPeripheral )
+{
+    int32_t lError = IOT_SPI_SUCCESS;
+
+    if( ( pxSPIPeripheral == NULL ) || ( pxSPIPeripheral->sOpened == IOT_SPI_CLOSED ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        if( HAL_SPI_Abort( pxSpi ) != HAL_OK )
+        {
+            lError = IOT_SPI_BUS_BUSY;
+        }
+        else if( HAL_SPI_DeInit( pxSpi ) != HAL_OK )
+        {
+            lError = IOT_SPI_BUS_BUSY;
+        }
+        else
+        {
+            pxSPIPeripheral->sOpened = IOT_SPI_CLOSED;
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_spi_cancel( IotSPIHandle_t const pxSPIPeripheral )
+{
+    int32_t lError = IOT_SPI_INVALID_VALUE;
+
+    if( ( pxSPIPeripheral != NULL ) && ( pxSPIPeripheral->sOpened == IOT_SPI_OPENED ) )
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        if( HAL_SPI_GetState( pxSpi ) == HAL_SPI_STATE_READY )
+        {
+            lError = IOT_SPI_NOTHING_TO_CANCEL;
+        }
+        else if( HAL_SPI_Abort( pxSpi ) == HAL_OK )
+        {
+            lError = IOT_SPI_SUCCESS;
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+void HAL_SPI_TxRxCpltCallback( SPI_HandleTypeDef * hspi )
+{
+    if( hspi->Instance == SPI1 )
```

Comment:
Rather than require these if/else blocks, it seems like the respective user context pointer should be in the SPI instance handle structure.

Reply:
Ideally yes. SPI_HandleTypeDef structure is of ST driver's HAL while xSpiCallback is defined by us. I didn't find a good map of mapping SPI_HandleTypeDef pointer to IotSPIHandle_t
For now, let's keep it as is.